Fix crash and warnings when opening an iOS/arm64e DSC

oleavr opened the pull request:
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
Running r2 on the `dyld_shared_cache_arm64e` from my iOS/arm64e device running iOS 13.5, I ran into the following issues:
1. It crashes because the `THREADED` opcode is encountered before the usually preceding `SET_SEGMENT_AND_OFFSET_ULEB` opcode. Apple's state machinery handles this case.
2. The reason 1) happens is because parsing doesn't stop when encountering the `DONE` opcode, so it goes ahead and parses “junk”.
3. With 1) and 2) out of the way, a lot of warnings were printed to the console: `WARNING: r_meta_set_with_subtype: assertion 'm && size' failed (line 185)`. This happens because of [this](https://github.com/radareorg/radare2/blob/f255a82a1bb45614a921e3cf156ac37b2b40811d/libr/core/cbin.c#L2206-L2209) logic in `bin_symbols`, which attempts to add an `R_META_TYPE_COMMENT` using a symbol where `symbol->size == 0`.

I've solved these three issues in two separate commits:

1. Improve handling of Mach-O bind opcodes
    * Reset state between regular binds, lazy binds, and weak binds.
    * Stop parsing when encountering the DONE opcode, except for when
      processing lazy binds.
    * Rework handling of THREADED opcode to match Apple's behavior, where
      the segment index isn't used. Also fix the page_count calculation.
    * Eliminate redundant boundary checks to improve clarity.

2. Fix symbol comment logic for zero-sized symbols

**Test plan**
```sh
$ R_DYLDCACHE_FILTER=OSAnalytics r2 dyld_shared_cache_arm64e-ios-13.5-iphone-xsmax
```

**Closing issues**
None that I'm aware of.

oleavr commented:
Seeing this test failure:

```
[XX] db/formats/mach0/thumb mach0 non-thumb symbol on thumb bin
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'pi 1@ sym.imp.strcmp' bins/mach0/thumb-test
-- stdout
--- .a	2020-05-27 23:01:32.987922284 +0000
+++ .b	2020-05-27 23:01:32.987922284 +0000
@@ -1 +1 @@
-ldr pc, [reloc.strcmp]
+ldr pc, [segment.DATA]
```

I'm wondering if my assumption about `BIND_OPCODE_DONE` is wrong?

/cc @mrmacete

XVilka commented:
Still red:
```
[**]                              db/formats/mach0/objc     1461 OK        28 BR        0 XX        1 FX


[XX] db/formats/mach0/swift mach0 swift-x86-64 aav
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'C*~Cd?
aav
C*~Cd?
Cd~?
' bins/mach0/swift-main
-- stdout
--- .a	2020-05-28 03:38:01.808797816 +0000
+++ .b	2020-05-28 03:38:01.808797816 +0000
@@ -1,3 +1,3 @@
-114
-134
+115
+135
134


-- stderr
[ ] Finding xrefs in noncode section with anal.in=io.maps
```

oleavr commented:
@XVilka Thanks! Investigating.

oleavr commented:
@XVilka Looks like the test got broken by the `bin_symbols()` introduction of `R_META_TYPE_COMMENT`, but this was hidden because of the bug that I fixed here causing: `WARNING: r_meta_set_with_subtype: assertion 'm && size' failed (line 185)`. Will update the test and take it for a new spin on the CI.

Here's the diff between `C*~Cd` before and after the fix for 3):

```diff
--- before	2020-05-28 05:55:42.000000000 +0200
+++ after	2020-05-28 05:54:55.000000000 +0200
@@ -1,3 +1,4 @@
+CCu base64:U3dpZnQuU3RyaW5nLmluaXQoX2J1aWx0aW5TdHJpbmdMaXRlcmFsOiBCdWlsdGluLlJhd1BvaW50ZXIsIGJ5dGVTaXplOiBCdWlsdGluLldvcmQsIGlzQVNDSUk6IEJ1aWx0aW4uSW50MSkgLT4gU3dpZnQuU3RyaW5n @ 0x10000440a
 Cd 8 @ 0x100005000
 Cd 8 @ 0x100005010
 Cd 8 @ 0x100005018
```

XVilka commented:
@mrmacete do you approve this PR?

ret2libc commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-			switch (op) {
+		int partition_sizes[] = {bind_size, lazy_size, weak_size};
+		int opcodes_offset = 0;
+		for (int partition_idx = 0; partition_idx < R_ARRAY_SIZE (partition_sizes); partition_idx++) {
```
do not define this inside the for-loop. We don't use `std=c99` for these files AFAIK, so it would not work on some compilers.

oleavr replied:
Oops, great catch! Fixed.

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					break;
+				case BIND_OPCODE_SET_SYMBOL_TRAILING_FLAGS_IMM: {
+					sym_name = (char*)ur.p;
+					while (*ur.p++ && ur.p<end) {
```
spaces around <

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-							if (sidx < 0 || sidx >= bin->nsymtab) {
-								continue;
+				case BIND_OPCODE_SET_DYLIB_ORDINAL_ULEB:
+					lib_ord = ULEB();
```
missing space before (

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					if (seg_idx >= bin->nsegs) {
+						bprintf ("Error: BIND_OPCODE_SET_SEGMENT_AND_OFFSET_ULEB"
+							" has unexistent segment %d\n", seg_idx);
+						R_FREE (opcodes);
```
just use free, because there's a return nearby and there's no need to nullify the pointer

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3425,56 +3432,60 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-					if (addr >= segmentAddress) {
-						bprintf ("Error: Malformed ULEB TIMES bind opcode\n");
+				case BIND_OPCODE_DO_BIND:
+					if (!threaded_binds && addr >= segment_end_addr) {
```
imho the second condition is >, not >=

oleavr replied:
The end address is exclusive, so `>=` means it will reject addresses outside the segment. Does it make sense to allow a bind address that's pointing at the first byte after the segment?

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3425,56 +3432,60 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+						bprintf ("Error: Malformed IMM SCALED bind opcode\n");
+						goto beach;
+					}
+					DO_BIND();
```
missing space

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3425,56 +3432,60 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					addr += (ut64)imm * (ut64)wordsize + wordsize;
+					break;
+				case BIND_OPCODE_DO_BIND_ULEB_TIMES_SKIPPING_ULEB:
+					count = ULEB();
```
missing space  same in the line below

radare commented on libr/core/cbin.c:
```diff
@@ -2205,7 +2205,7 @@ static int bin_symbols(RCore *r, int mode, ut64 laddr, int va, ut64 at, const ch
 			if (sn.demname) {
 				r_meta_set (r->anal, R_META_TYPE_COMMENT,
-							addr, symbol->size, sn.demname);
+							addr, symbol->size? symbol->size: 1, sn.demname);
```
imho is better to have a separate variable instead of making that line that long.

```
const size_t sym_size = symbol->size > 0? symbol->size: 1;
```

radare commented on test/db/formats/mach0/swift:
```diff
@@ -18,8 +18,8 @@ C*~Cd?
 EXPECT=<<EOF
-114
-134
+115
```
which is the line that is new? can we have a test that provides an insight of the fix instead of just a row count?

oleavr replied:
Previously all lines looked like this:

```
Cd 8 @ 0x100005000
Cd 8 @ 0x100005010
Cd 8 @ 0x100005018
Cd 8 @ 0x100005020
```

Now there's also this line:

```
CCu base64:U3dpZnQuU3RyaW5nLmluaXQoX2J1aWx0aW5TdHJpbmdMaXRlcmFsOiBCdWlsdGluLlJhd1BvaW50ZXIsIGJ5dGVTaXplOiBCdWlsdGluLldvcmQsIGlzQVNDSUk6IEJ1aWx0aW4uSW50MSkgLT4gU3dpZnQuU3RyaW5n @ 0x10000440a
```

So in hindsight we should probably instead make the "grep" pattern stricter so it doesn't result in false positives when `Cd` just happens to be in an unrelated line's base64-encoded data.

oleavr replied:
Fixed it by adding `^` to make this test less brittle.

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+						r_skiplist_free (relocs);
+						r_pvector_free (threaded_binds);
+						return NULL; // early exit to avoid future mayhem
+					} else {
```
don't else after return

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-				break;
-			}
-			switch (op) {
+		int partition_sizes[] = {bind_size, lazy_size, weak_size}, partition_idx;
```
partition_idx must be size_t because it cant be negative

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -1591,7 +1590,7 @@ static bool reconstruct_chained_fixup(struct MACH0_(obj_t) *bin) {
 						cur_seg->pointer_format = DYLD_CHAINED_PTR_ARM64E;
 						cur_seg->page_size = 0x1000;
-						cur_seg->page_count = (bin->segs[seg_idx].vmsize + (0x1000 - 1)) & ~(0x1000 - 1);
+						cur_seg->page_count = ((bin->segs[seg_idx].vmsize + (0x1000 - 1)) & ~(0x1000 - 1)) / 0x1000;
```
maybe better to reuse the pagesize instead of hardcoding the 4KB (0x1000) 3 times. you can defne a variable above like this:

`const size_t ps = 0x1000;` and use `ps` (from `pagesize`) (just trying to make it shorter)

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3159,16 +3158,10 @@ static void parse_relocation_info(struct MACH0_(obj_t) *bin, RSkipList * relocs,
-		st64 addend = 0;
-		ut64 segmentAddress = 0LL;
-		ut64 addr = 0LL;
+		ut8 *opcodes, rel_type = 0;
```
it's a pointer, so it must use NULL, not 0

oleavr replied:
Only `opcodes` is, `rel_type` is a `ut8`.

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-			}
-			switch (op) {
+		int partition_sizes[] = {bind_size, lazy_size, weak_size}, partition_idx;
+		int opcodes_offset = 0;
```
this variable is never modified. can be removed

oleavr replied:
It's modified at the end of the loop.

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-			switch (op) {
+		int partition_sizes[] = {bind_size, lazy_size, weak_size}, partition_idx;
+		int opcodes_offset = 0;
+		for (partition_idx = 0; partition_idx < R_ARRAY_SIZE (partition_sizes); partition_idx++) {
```
i prefer `pidx` as its shorter to read as an iterator variable

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+		int partition_sizes[] = {bind_size, lazy_size, weak_size}, partition_idx;
+		int opcodes_offset = 0;
+		for (partition_idx = 0; partition_idx < R_ARRAY_SIZE (partition_sizes); partition_idx++) {
+			int partition_size = partition_sizes[partition_idx];
```
must be size_t

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+			bool done = false;
+			while (!done && ur.p < end) {
+				ut8 imm = *ur.p & BIND_IMMEDIATE_MASK, op = *ur.p & BIND_OPCODE_MASK;
+				++ur.p;
```
do not preincrement if its not necessary. `ur.p++`

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+			ut8 *end = ur.p + partition_size;
+			bool done = false;
+			while (!done && ur.p < end) {
+				ut8 imm = *ur.p & BIND_IMMEDIATE_MASK, op = *ur.p & BIND_OPCODE_MASK;
```
better use two separated variable definitions. it ill be cleaner to read (s/,/;\n/)

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					switch (imm) {
+					case BIND_SUBOPCODE_THREADED_SET_BIND_ORDINAL_TABLE_SIZE_ULEB: {
+						ut64 table_size = ULEB();
+						if (table_size == 0 || table_size > 65535) {
```
use constants or move this logic into a separate function to make the code more readable (`if (is_valid (table_size)) {`). Also use UT16_MAX instead of 65535

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+				case BIND_OPCODE_THREADED: {
+					switch (imm) {
+					case BIND_SUBOPCODE_THREADED_SET_BIND_ORDINAL_TABLE_SIZE_ULEB: {
+						ut64 table_size = ULEB();
```
missing space

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					case BIND_SUBOPCODE_THREADED_APPLY:
+						if (threaded_binds) {
+							int cur_seg_idx = (seg_idx != -1)? seg_idx: 0;
 							int n_threaded_binds = (int) r_pvector_len (threaded_binds);
```
this cant be negative, must be size_t

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3227,44 +3220,61 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					}
+					case BIND_SUBOPCODE_THREADED_APPLY:
+						if (threaded_binds) {
+							int cur_seg_idx = (seg_idx != -1)? seg_idx: 0;
```
this is wrong. seg_idx cant be -1 after your change. but imho the wronginess is in not initializing that index with -1

oleavr replied:
It can here, because here we start out with `-1` like Apple's state machine does. This is used for error-checking for the non-threaded opcodes. But for `reconstruct_chained_fixup()` we don't care about those non-threaded opcodes, so there we initialize it to `0`.

This is a bit confusing because Apple's state machine requires setting the current segment, except for when handling the threaded opcodes – there they don't care about the index and use the segment start + end, which are initially set based on the start/end of the segment with index 0.

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					sym_ord = -1;
+					if (bin->symtab && bin->dysymtab.nundefsym < 0xffff) {
+						for (j = 0; j < bin->dysymtab.nundefsym; j++) {
+							int stridx = 0;
```
maybe initialize with -1 in the case the symbol is not found it will be using the first one (index 0), which is imho wrong

radare commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
-							stridx = bin->symtab[sidx].n_strx;
-							if (stridx < 0 || stridx >= bin->symstrlen) {
-								continue;
+							if (!strcmp ((char *)bin->symstr + stridx, sym_name)) {
```
the cast must be const char * ,not char *

oleavr commented on libr/bin/format/mach0/mach0.c:
```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
+					sym_ord = -1;
+					if (bin->symtab && bin->dysymtab.nundefsym < 0xffff) {
+						for (j = 0; j < bin->dysymtab.nundefsym; j++) {
+							int stridx = 0;
```
Note that I only changed the indentation level of this code to keep the scope as narrow as possible – but if we're changing these lines we might as well audit them 👌

```C
int stridx = 0;
int iundefsym = bin->dysymtab.iundefsym;
if (iundefsym >= 0 && iundefsym < bin->nsymtab) {
	int sidx = iundefsym + j;
	if (sidx < 0 || sidx >= bin->nsymtab) {
		continue;
	}
	stridx = bin->symtab[sidx].n_strx;
	if (stridx < 0 || stridx >= bin->symstrlen) {
		continue;
	}
}
if (!strcmp ((char *)bin->symstr + stridx, sym_name)) {
	sym_ord = j;
	break;
}
```

So `stridx` is declared and initialized to zero, but never read until it's assigned a new value here:

```C
stridx = bin->symtab[sidx].n_strx;
```

Here, `n_strx` is a `uint32_t`. And on the next line we check:

```C
	if (stridx < 0 || stridx >= bin->symstrlen) {
```

Where `stridx < 0` is impossible, so that should be removed. So we're left with this check:

```C
	if (stridx >= bin->symstrlen) {
```

Where `symstrlen` is an `int`. So ISTM we caan just drop the variable and inline that into the check:

```C
if (bin->symtab[sidx].n_strx >= bin->symstrlen) {
```

I also noticed that `iundefsym` is assigned to an `int` here, also coming from an `uint32_t`, and checking if it's `>= 0` (anything else is impossible), and used to compute `sidx`, also an `int` that cannot be `< 0`, but still there's a check for that.

oleavr replied:
Ignore that, `ENOCOFFEE` so I totally misread that code 😂

oleavr commented:
@radare Thanks! Believe I've addressed all of the issues except for adding a test.

Test coverage for this seems like it will be really hard as the `dyld_shared_cache` that I reproduced this on is 1.6 GB and not redistributable. I've started a side project with [dyld-tools](https://github.com/oleavr/dyld-tools) that I hope will make it possible to generate a miniature `dyld_shared_cache` that can be added to radare2-testbins and used for writing tests.

I don't know whether it's worth keeping this PR in limbo until I get to that point though, as it might take a long time to get there.

oleavr commented:
Seems the refactoring may have broken this test:

```
[XX] db/anal/classes anal classes arm64e
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'avrr
acl
acll~\@
avra~0x100008370:0
avra~vmi:0
?e `avraj~{[0]}`~{:
' bins/mach0/TestRTTI-arm64e
-- stdout
--- .a	2020-06-01 18:10:30.279451134 +0000
+++ .b	2020-06-01 18:10:30.279451134 +0000
@@ -15,7 +15,6 @@
virtual_8 @ 0x1000060a0 (vtable + 0x8)
virtual_16 @ 0x10000615c (vtable + 0x10)
Type Info at 0x100008370:
-  Type Info type: __vmi_class_type_info
type: __class_type_info
found_at: 4295000200
class_vtable: 4295000176
```

@mrmacete Anything obvious that I screwed up that could explain this? 😅

XVilka commented:
Because adding test is not so easy, I suggest we merge this PR without waiting for them, since it fixes the crash.

mrmacete commented:
LGTM
